SEP-1686: Tasks

LucaButBoring opened the pull request:
This PR implements the required changes for https://github.com/modelcontextprotocol/modelcontextprotocol/issues/1686, which adds task augmentation to requests.

## Motivation and Context
The current MCP specification supports tool calls that execute a request and eventually receive a response, and tool calls can be passed a progress token to integrate with MCP’s progress-tracking functionality, enabling host applications to receive status updates for a tool call via notifications. However, there is no way for a client to explicitly request the status of a tool call, resulting in states where it is possible for a tool call to have been dropped on the server, and it is unknown if a response or a notification may ever arrive. Similarly, there is no way for a client to explicitly retrieve the result of a tool call after it has completed — if the result was dropped, clients must call the tool again, which is undesirable for tools expected to take minutes or more. This is particularly relevant for MCP servers abstracting existing workflow-based APIs, such as AWS Step Functions, Workflows for Google Cloud, or APIs representing CI/CD pipelines, among other applications.

This proposal (and implementation) solves this by introducing the concept of Tasks, which are pending work objects that can augment any other request. Clients generate a task ID and augment their request with it — that task ID is both a reference to the request and an idempotency token. If the server accepts the task augmentation request, clients can then poll for the status and eventual result of the task with the `tasks/get` and `tasks/result` operations.

## How Has This Been Tested?
Unit tests and updated sHTTP example.

## Breaking Changes
None.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

halter73 commented on src/shared/protocol.ts:
```diff
@@ -360,31 +510,100 @@ export abstract class Protocol<SendRequestT extends Request, SendNotificationT e
+                    return await this.request(r, resultSchema, { ...options, relatedRequestId: request.id, relatedTask });
+                } finally {
+                    if (taskMetadata && this._taskStore) {
+                        await this._taskStore.updateTaskStatus(taskMetadata.taskId, 'working');
```
This also seems like it could race. While hopefully a request handler never completes before a request it makes, timeouts and the like make this inevitable.

If the expectation is that the TaskStore should be responsible for dealing with this race, that's all the more reason to include a production-ready in-memory TaskStore in the box that makes it clear what the expectations are for a TaskStore to be production-ready.

halter73 commented on src/shared/request.ts:
```diff
@@ -0,0 +1,75 @@
+import { Request, Notification, Result, Task, GetTaskResult } from '../types.js';
+import { isTerminal } from './task.js';
+
+const DEFAULT_POLLING_INTERNAL = 5000;
```
Was this meant to be `DEFAULT_POLLING_INTERVAL`?

And speaking of internal, shouldn't this be configurable? It also might make sense to allow people to override the polling interval on the client-side and disregard what the server suggests since respecting it is only a SHOULD in the SEP, but at the very least you should be able to configure the default when the server doesn't specify anything.

halter73 reviewed (commented):
SEP-1686 currently states that it "introduces a mechanism for requestors (which can be either clients or servers, depending on the direction of communication) to augment their requests with tasks." Is this still the case?

I don't see any tests or examples demonstrating using a TaskStore with an McpClient (it's all McpServer or the Protocol base type), although I suppose it should work considering its shared code. It still might be nice to have an end-to-end test demonstrating that client-side support for stuff like elicitations works using the public APIs.

LucaButBoring commented on src/shared/protocol.ts:
```diff
@@ -211,19 +256,105 @@ export abstract class Protocol<SendRequestT extends Request, SendNotificationT e
+        const taskId = this._requestIdToTaskId.get(Number(notification.params.requestId));
+        if (taskId && this._taskStore) {
+            try {
+                await this._taskStore.updateTaskStatus(taskId, 'cancelled');
```
That is a good callout, will fix this.

LucaButBoring commented:
> It still might be nice to have an end-to-end test demonstrating that client-side support for stuff like elicitations works using the public APIs.

Agreed, will add this.

He-Pin commented on src/server/index.ts:
```diff
@@ -271,13 +277,74 @@ export class Server<
+            case 'tasks/get':
+            case 'tasks/list':
+            case 'tasks/result':
+            case 'tasks/delete':
```
tasks/cancel

He-Pin commented on src/shared/task.ts:
On the client side from my current code: 
```java
@InternalStableApi
sealed public interface McpTaskProtocol {
    sealed interface McpTaskCommand extends McpTaskProtocol {

        /**
         * 有新的任务需要开始轮询
         */
        // @formatter:off
        record StartTaskPollingCommand(
            TmcpAsyncClient tmcpAsyncClient, //tmcp客户端, 用于发起轮询请求
            TmcpClientSession session, //session
            SynchronousSink<McpSchema.CallToolResult> resultSink, //最终的结果sink
            String requestId, //发起任务请求的id
            String methodName, //目前只会是 tools/call
            McpSchema.Task task, //任务信息
            Map<String, Object> metadata //任务相关的元数据
            ) implements McpTaskCommand {
        }
        // @formatter:on
    }

    sealed interface McpTaskEvent extends McpTaskProtocol {
        /**
         * 通过 `tasks/get`, `tasks/list` , `tasks/cancel` 等接口拉取到了任务的最新状态
         * <p>
         * 我们将会把这个状态提交给任务的观察者们
         */
        record TaskStatusUpdatedEvent(McpSchema.TaskInfo task, Map<String, Object> meta) implements McpTaskEvent {
        }

        /**
         * 通过 `tools/result` 接口拉取到了任务的最终结果
         * */
        record TaskResultRetrievedEvent(String taskId, McpSchema.CallToolResult result) implements McpTaskEvent {
        }
    }
}

```

I'm currently handle it with `event bus`'s pub sub model, and let the implementation detail as how to handle the events.

LucaButBoring replied:
Expanded `QueuedMessage` to also include responses and errors in the latest commit.

LucaButBoring commented on src/examples/shared/inMemoryTaskStore.ts:
```diff
@@ -198,3 +198,82 @@ export class InMemoryTaskStore implements TaskStore {
+     * @param maxSize Optional maximum queue size - if specified and queue is full, throws an error
+     * @throws Error if maxSize is specified and would be exceeded
+     */
+    async enqueue(taskId: string, message: QueuedMessage, sessionId?: string, maxSize?: number): Promise<void> {
```
That's the intention behind making this an example implementation for the main interface, so consumers can satisfy the same interface with an event bus if they wish. Is there anything else you'd need here?

He-Pin commented on src/shared/task.ts:
Maybe `PendingTaskAction` ?

He-Pin commented:
I understand now, that makes a lot clearer. So this isn't a Queue; I think it's more of a Stash and Unstash logic.

LucaButBoring commented:
> Since I can use a distributed queue, why not just implement the transport between the server and client based on the distributed queue?

Can you clarify what you mean by this part?

If you mean that you would put the queue directly between the client and server, rather than having it _within_ the server, that sounds like a custom transport, if I'm interpreting it correctly. A custom transport that handles this would work, but we do need something for sHTTP, and I'd prefer not to make the sHTTP transport have special logic for particular JSON-RPC requests if that can be avoided. The message queue here is transport-agnostic.

Specifically, if this were pushed into the transport, sHTTP would need to handle this by:
1. Checking `related-task` to determine if it needs to queue a message to the client
2. Checking if a request is a `tasks/result` request, and if so, dequeueing all messages into the response stream

Both of these things sound like they should be the responsibility of `Protocol`, to me. `Protocol` is aware of there being request/response streams, but doesn't know what the transport is — likewise, the transport knows how the request/response streams are implemented, but doesn't know how the underlying protocol operations need to be handled (with the exception of `initialize`, unfortunately).

felixweinberger commented on src/client/index.ts:
```diff
@@ -527,64 +597,158 @@ export class Client<
-    ) {
-        const result = await this.request({ method: 'tools/call', params }, resultSchema, options);
+    ): Promise<SchemaOutput<T>> {
+        return await takeResult(this.callToolStream<T>(params, resultSchema, options));
```
elegant unification

felixweinberger commented on src/client/index.ts:
```diff
@@ -527,64 +597,158 @@ export class Client<
+     * @param options - Optional request options (timeout, signal, task creation params, etc.)
+     * @returns AsyncGenerator that yields ResponseMessage objects
+     */
+    async *callToolStream<T extends typeof CallToolResultSchema | typeof CompatibilityCallToolResultSchema>(
```
do we expect end users to actually call this function?

If so probably worth having it in the README.

felixweinberger commented on src/examples/server/simpleStreamableHttp.ts:
```diff
@@ -440,6 +455,65 @@ const getServer = () => {
+                // Create the task
+                const task = await taskStore.createTask(
+                    {
+                        taskId,
```
this is ignored afaict
```suggestion
```

felixweinberger commented on src/examples/server/simpleStreamableHttp.ts:
```diff
@@ -440,6 +455,65 @@ const getServer = () => {
+            }
+        },
+        {
+            async createTask({ duration }, { taskStore, taskRequestedTtl }) {
```
Couple of fixes were needed to make these examples work for me. As-is I got this:

<img width="1684" height="814" alt="Image" src="https://github.com/user-attachments/assets/1f520564-fe22-4c30-90d4-7f6f40c7ce7e" />
(note the error message on server on left about wrong taskId)

But with changes below I got:

<img width="1688" height="824" alt="Image" src="https://github.com/user-attachments/assets/d07ab65a-e73f-42fe-ac55-8e36d1402616" />

LucaButBoring replied:
Oh, actually, it looks like `delay` calls `createTask` with a 100ms interval explicitly, I think I did that for testing a while back - will adjust this to pass through the suggested polling interval.

LucaButBoring commented on README.md:
```diff
@@ -1382,6 +1383,175 @@ const client = new Client(
+
+// Initiate a task-based tool call
+const taskId = 'analysis-task-123';
+const pendingRequest = client.beginCallTool(
```
Also forgot to update this, it should actually be `callToolStream` in this particular example now, since I removed `PendingRequest` and `beginCallTool` last week.

LucaButBoring commented:
@felixweinberger Updated the example and README, let me know what you think 😄 

I also simplified the `RequestTaskStore` interface a bit more by removing the (also now-redundant) `requestId` and `request` parameters, since the `RequestTaskStore` shim automatically passes those to `TaskStore`. I missed a few things like that when I first created that wrapper and moved common logic into it, but I think I got everything now.

Finally, I added `lastUpdatedAt` to `Task`, which is the final revision accepted to the spec (we're doing a hard cutoff here): https://github.com/modelcontextprotocol/modelcontextprotocol/pull/1875

maxisbey commented on src/client/index.test.ts:
```diff
@@ -1690,6 +1785,1304 @@ describe('outputSchema validation', () => {
+            // Get the task ID from the task list since it's generated automatically
+            const taskList = await server.listTasks();
+            expect(taskList.tasks.length).toBeGreaterThan(0);
+            const taskId = taskList.tasks[0].taskId;
```
Instead of getting the task ID by getting the 0th item in the list of all tasks, could you instead get it via the CreateTaskResult object (same with all the above tests which bypass this in a similar way)

felixweinberger commented on src/shared/protocol.ts:
```diff
@@ -225,6 +361,186 @@ export abstract class Protocol<SendRequestT extends Request, SendNotificationT e
+                        let queuedMessage: QueuedMessage | undefined;
+                        while ((queuedMessage = await this._taskMessageQueue.dequeue(taskId, extra.sessionId))) {
+                            // Handle response and error messages by routing them to the appropriate resolver
+                            if (queuedMessage.type === 'response' || queuedMessage.type === 'error') {
```
Tracing through at the moment, where do `'response'` and `'error'` type messages ever get enqueued?

The tests only seem to directly enqueue the messages to "simulate a client response" - but where does this capability actually live on the client at the moment?

LucaButBoring replied:
Putting this here for reference as it's the intended flow: [Mermaid](https://mermaid.live/edit#pako:eNq1V99z4jYQ_lc0fggwQzgIiQl-uJmWpqVzbS_EfmrpZFR7Qzwxkk-Sk0sz_O9dWbJjbOOQ6R0vgPm02h_ffru8OCGPwPEcCV8yYCH8FNONoNs1I_hKqVBxGKeUKbIgVJJFEgN-vhZc8ZAnLaigAgsEZTLlQjVxfo7zQTyC6MRVYIdvDVYaFlD58DtISTewyiCDFpxf4HzFRQtgqX__QT6zkCwpixIQa2ZQf3AFhGs3FsOlR67vqQQy8XJjZCGAqphbKA1V_EgRvjDfF6cfPy4CjyjOE_khpElCXojCc55-V4lH3DG-yI7s6gYCayFAEz6aWAbBNbn-7Ac1YPHAz4EeueVM6IJK1R_UoRaJwACRoXYddBQNZGChgX9qrOaxGtd_jYZEKqoy6ZHeExcPMdv0Cv8jaNhAEwNMm68w2ZhjneB7k2DSZ5yd_pPwUNuo-7Asj5sELEp3b0BmibLutNzslxmx6c9zh6dSziQ04a-p1vjmTY0TRXGqjzrZcuYVTM7pKclVgrxTOXPQsbxc5vjS5Bu-KkFHElhkf-3D64khhi4gwWujvC67rkKv0Bj7om_tY8KeU8CqWYL0hmRrmgYxFYca9oKVpcPK0uHzp2a5V6-X-rrc2GXopeSJDjtm5NbeemOfyXq-8BT2-TbGfD3RWCElRqPRoeIuLYmzNLK18nNO9nsxSzOVXxYLiHrd1F62x-LXfUPcj5qnEBFqndtLmSjJ1cGCqVeqKE8SSYzHbygHRiY_bECV3fddpOIXUDqHbyjGxqD6xpFjRaMUi1plujXDhKJNfPP-fr9f727487LUP4MK77Hj88aPiJ1SR1VdVHXu-xb-mj4nnEZvCEkERkgO1L-uEkfrTbeYYGBaCA8ooNVHnZ5BnRe-f0WuHrEKXq1ZK4K7pz928pN8JkmSB0oYV0SB2MaMJoPOql-UVTeGJDmxtIxkpVSLvQpUPBu0s6Jyl0eu9GQIVXUAYOx3KJzkdguKVoiE3uiNw8wamycfybilxQTeB99DiIvJvuGTE3KrWrerPa8aGBJLkrEI7mIGUY3cxc6Xz7d-IZ2DQ2Qm_TalHbyzI91yBN9ACPEjFqdQjCGxM0kWA6ilXQy0yxU7-X7j_CFLjxl99kCBrGeidTJasKzoJIZ4te-VUQ2dGM6w_WgYQqq7L-RMaXbuWsabXnWwO3HzgDAzG21HNmfF-su3aQKq8KeYylJP_9cFqt8LLS5CL-50I5kf_s9wrjxatiRsX9jKlfOJPmCQWfpNB9trdJ0zbe8um5kjb9SNHOCfiAMLaWNu3oDKBDOptox4e4xW5PKN-5or8x4ec2Op5pG_cIn7-8iR6gydjYgjx1Mig6GzRdGl-qvzorFrR6F2wdrx8GNExcPaWbMdnsH_b39yvi2OCZ5t7h3vjibY2o5ZEO1_2_KpQOUBseAZU443m7mXuRXHe3G-Ot504o5m0_nMnbljdz6-mLtD59nx3PPRdDp152fT-WQ-n44vdkPn3_ze8Wh-OTm7mLgXs_F4Pjk7v9z9B86k9n0)
